test/functional/feature_index_prune.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test indices in conjunction with prune."""
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_greater_than,
+    assert_raises_rpc_error,
+    p2p_port,
+)
+
+
+class FeatureIndexPruneTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 4
+        self.extra_args = [
+            ["-fastprune", "-prune=1", "-blockfilterindex=1"],
+            ["-fastprune", "-prune=1", "-coinstatsindex=1"],
+            ["-fastprune", "-prune=1", "-blockfilterindex=1", "-coinstatsindex=1"],
+            []
+        ]
+
+    def sync_index(self, height):
+        expected_filter = {
+            'basic block filter index': {'synced': True, 'best_block_height': height},
+        }
+        self.wait_until(lambda: self.nodes[0].getindexinfo() == expected_filter)
+
+        expected_stats = {
+            'coinstatsindex': {'synced': True, 'best_block_height': height}
+        }
+        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
+
+        expected = {**expected_filter, **expected_stats}
+        self.wait_until(lambda: self.nodes[2].getindexinfo() == expected)
+
+    def reconnect_nodes(self):
+        self.connect_nodes(0,1)
+        self.connect_nodes(0,2)
+        self.connect_nodes(0,3)
+
+    def mine_batches(self, n):
+        for _ in range(n):
+            self.nodes[0].generate(250)
+            self.sync_blocks()
+
+    def expected_prune_height(self, n):
+        # Sanity check for correct prune blocker height in logs
+        for node in self.nodes:
+            node.assert_debug_log([f'limited pruning to height {n}'])
+
+    def run_test(self):
+        filter_nodes = [self.nodes[0], self.nodes[2]]
+        stats_nodes = [self.nodes[1], self.nodes[2]]
+
+        self.log.info("check if we can access blockfilters and coinstats when pruning is enabled but no blocks are actually pruned")
+        self.sync_index(height=200)
+        tip = self.nodes[0].getbestblockhash()
+        for node in filter_nodes:
+            assert_greater_than(len(node.getblockfilter(tip)['filter']), 0)
+        for node in stats_nodes:
+            assert(node.gettxoutsetinfo(hash_type="muhash", hash_or_height=tip)['muhash'])
+
+        # Mine two batches of blocks to avoid hitting NODE_NETWORK_LIMITED_MIN_BLOCKS disconnection
+        self.mine_batches(2)
+        self.sync_index(height=700)
+
+        self.expected_prune_height(690)
```

Comment:
I am able to put any value here besides 690 and the test still passes.

This is because [`assert_debug_log`](https://github.com/bitcoin/bitcoin/blob/5c0f46ca46e23a161649b5150baa01020dc85e48/test/functional/test_framework/test_node.py#L381) is a generator and meant to be used in combination with `with`:
```python
with assert_debug_log(...):
  # some other code that prints the (un)expected logs
```
"At the point where the generator yields, the block nested in the [`with`](https://docs.python.org/3/reference/compound_stmts.html#with) statement is executed. The generator is then resumed after the block is exited. [...]" from https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

Secondly i think that checking the logs to test the prune blocker height won't work here as the expected message is only printed when pruning is attempted (https://github.com/bitcoin/bitcoin/blob/983a3552f22678d4466e0faa8e784f25b145a76b/src/validation.cpp#L2027)

the following would work:
```python
 self.log.info("prune some blocks")
 for node in self.nodes[:2]:
     with node.assert_debug_log(['limited pruning to height 690']):
         pruneheight = node.pruneblockchain(400)
         assert_equal(pruneheight, 248)
```

Reply:
Thanks a lot for the detailed feedback! I fixed this using your suggested approach in the two locations where I wanted to do the check.